[Ang Yong Liang] iP

yl-ang opened the pull request:
# iP milestone 1
## Recent updates
- **Junit tests**
- *Error handling*
- **Java doc**

## Future updates
1.  GUI
2. Varargs
3. Additional unit tests

## Junit test for Storage class :+1:

```java
public class storageTest {

    @Test
    public void loadTest() {
        assertEquals("[T][ ] coding", new Storage("data/tasks.txt").load().get(0).toString());
    }

}
```
`
Test 
`
## Todo
- [x] A-Gradle
- [x] A-CheckStyle
- [x] Level-10
- [ ] A-Varargs

> ~~Text that is a quote~~

This site was built using [GitHub Pages](https://pages.github.com/).

Commit:
Add Gradle support

yl-ang committed:
Update Duke.java

Level-1: Greet, Echo, Exit
* created printLine()
* created greet()
* created echo()
* created exit()

yl-ang committed:
Updated Duke.java

Level 2. Add, List
* Added a private arrayList attribute to Duke
* Added the ability to store whatever text entered by user
* Added the ability to display the stored text back to the user

yl-ang committed:
Finished Level-3

Level-3: Mark as Done

Changes to Duke.java
* modify and renamed lst attribute in duke to contain Task Object
* add indentation to text to be displayed
* create the retrieveTask method to retrieve Task from attribute
* create markOrUnmarked to mark or unmark tasks

New Task.java file
* add new Task class to support representation of tasks

yl-ang committed:
Finish Level-4

Improved Duke.java
* Removed unnecessary System.out.println() calls and
  depend more on echo().
* implemented the extractStr method to extract specific
  substrings from a string, to extract description and date.

Added 3 different types of Tasks
* Event, DeadLine and ToDo are subclasses of Task, using
  inheritance will later help us to make use of polymorphism
  when dealing with different types of Tasks in the arrayList.

yl-ang committed:
Finish A-TextUiTesting

* Included some input for text-ui-test
* Included expected ouput for FC
* Conducted automated testing of DUKE.
* Actual input matches expected output.

yl-ang committed:
Finish Level-5

* Created DukeException class
* Added try catch blocks to handle errors thrown by Duke

yl-ang committed:
Finish Level-6

* Created the delete task method.
* Continue to use ArrayList.

yl-ang committed:
Add error code via enums

* Fix bugs in extracting description and date
* Conducted additional automated testings

yl-ang committed:
Add more error handling

Added more error handlings such as
if integer is not provided to mark or unmark.

yl-ang committed:
Add error handle for empty delete

* error check for delete without any task number
* add errorString for empty delete into errorString enum

yl-ang committed:
Add more error checks

* error checks for already marked and already unmarked tasks
* error checks for empty deadline and event descriptions

yl-ang committed:
Finish Level-7

* Created save feature
* handled basic errors
* file path could be improved

yl-ang committed:
Fix save feature

* switch the directory from out to data/
* add File error handling
* improve command checks to prepare for packaging

yl-ang committed:
Improve save feature

* Add file dir and file exist checks

yl-ang committed:
Add minimal deadline date handling

* accepts dates in a format such as yyyy-mm-dd

yl-ang committed:
Merge branch 'branch-Level-7'

yl-ang committed:
Add date error handling for deadline

* check if the date input is of the format yyyy-mm-dd.

yl-ang committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/errorHandle/ErrorString.java

* Resolved conflict by accepting the ErrorString from
  branch-level-7

yl-ang committed:
Update gitignore

* ignore /data

yl-ang committed:
Add load feature

The load feature will allow the user to load back their existing
tasks from their duke.txt file.

yl-ang committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/task/DeadLine.java

yl-ang committed:
Remove wildcard import statement

yl-ang committed:
Make the code more OOP

Create classes to spilt up Duke.java to handle
different types of interactions and features.

yl-ang committed:
Fix delete feature

* Fix application unable to delete even with valid index.
* include more error handlings
* package classes

yl-ang committed:
Fix scanner bug

* Make scanner object a private attribute of UI
* Add more error checks

yl-ang committed:
Fix scanner bug

* Make scanner object a private attribute of UI
* Add more error checks

yl-ang committed:
Add JUnit Tests

* Junit test for ToDo and Storage classes

yl-ang committed:
Create a JAR file

Package the app as an executable JAR file so that it can
be distributed easily

yl-ang committed:
Update gitignore

ignore /data

yl-ang committed:
Add find command

* Basic find command without error handling

yl-ang committed:
Add error handling for find command

* consider IndexOutOfBoundsException when splitting user input

yl-ang committed:
Add java documentation

* Add java doc for public classes and methods.

yl-ang committed:
Merge branch 'branch-Level-9'

yl-ang committed:
Merge branch 'branch-A-JavaDoc'

yl-ang committed:
Follow coding standard

yl-ang committed:
Merge branch 'branch-A-CodingStandard'

Resolve conflicts

yl-ang committed:
Add class header comments for classes

yl-ang committed:
Fix find command

yl-ang committed:
Merge remote-tracking branch 'origin/add-gradle-support'

yl-ang committed:
Update JUnit test for Storage class

yl-ang committed:
Update build.gradle to run ShadowJar

yl-ang committed:
Merge branch 'branch-A-Gradle'

yl-ang committed:
Use CheckStyle

Add AB Check Style config files to test.

yl-ang committed:
Merge branch 'branch-A-CheckStyle'

yl-ang committed:
Update files to pass CheckStyle

yl-ang committed:
Add basic GUI for Duke

yl-ang committed:
Improve GUI

yl-ang committed:
Merge branch 'branch-Level-10'

yl-ang committed:
Refactor packages

yl-ang committed:
Add background picture to GUI

yl-ang committed:
Merge branch 'master' into branch-Level-10

yl-ang committed:
Improve Coding Standard

yl-ang committed:
Merge branch 'master' into branch-Level-10

yl-ang committed:
Fix background wallpaper for GUI

yl-ang committed:
Merge branch 'branch-Level-10'

Rye-Catcher commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     *
+     * @param filePath FilePath the path to the file containing the tasks.
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
```
The code looks incredibly good :thumbsup: . I just have a question: does this handle the file path on different OS?

yl-ang replied:
Thank you for the advice, I will test it on other OS (such as Mac) and check for consistency 👍 😄

Rye-Catcher commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,133 @@
+                throw new DukeException(ErrorString.ERROR_EMPTY_TODO_DESC.toString());
+            }
+
+            String desc = extractDesc(data, "todo ", null);
+            task = new ToDo(desc);
+
+        } else if (taskCommand.equals("deadline")) {
+
+            if ((temp.length == 1) || (temp[1].equals("/by"))) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_DEADLINE_DESC.toString());
+            }
+
+            if (!data.contains("/by")) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_DEADLINE_DATE.toString());
+            }
+
+            String desc = extractDesc(data, "deadline ", "/by");
+            String date = extractTime(data, "/by ");
+
+            task = new DeadLine(desc, date);
+
+        } else {
+
+            // Event handling
+
+            if ((temp.length == 1) || (temp[1].equals("/at"))) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_EVENT_DESC.toString());
+            }
+
+            if (!data.contains("/at")) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_EVENT_DATE.toString());
+            }
+
+            String desc = extractDesc(data, "event ", "/at");
+            String date = extractTime(data, "/at ");
+            task = new Event(desc, date);
+        }
+
+        this.task = task;
+    }
```
Clever work. This part handles exceptions in an elegant way

Rye-Catcher reviewed (commented):
Overall, the code quality is incredibly good :thumbsup:. 
Public methods are well documented. Naming and layouts are also correct. I do learn a lot from your code, like how to handle UI and exceptions. 
Btw the Ganyu picture is very cute.

yl-ang committed:
Fix UI text bug

yl-ang committed:
Merge branch 'branch-Level-10'

yl-ang committed:
Update build.gradle and Duke.java

The build.gradle file is outdated. In addition, there is no need
for the main and run methods in the current Duke class since we are
more focus on the GUI.

Updating build.gradle allows shadowJar to build a compatible jar
file for javaFx and updating Duke.java will improve code quality
via removable of dead code.

yl-ang committed:
Add assertions to Ui.java and other Command java class files

Assertions allows duke to terminate in the case of a possible bug,
and sooner the execution stops, the safer it is.

Multiple checks for null in the parameters passed into Ui.java and
the string response returned from command execution.

By checking for null will allow us to verify and ensure that the
program is behaving as expected.

yl-ang committed:
Merge pull request #1 from yl-ang/branch-A-Assertions

Add assertions to Ui.java and other Command java class files

yl-ang committed:
Merge branch 'master' of https://github.com/yl-ang/ip

yl-ang committed:
Improve code quality

Presence of dead code such as welcomeResponse() and lineResponse() in
Duke.java and unclear prominent path in certain classes.

Remove the dead code and introduce guard clauses into our program.

Removing dead code as we can simply recover it from revision control
tool when we need it in the future and use of guard clause in some code
will help us to make the code more reader friendly.

yl-ang committed:
Merge pull request #2 from yl-ang/branch-A-CodeQuality

Improve code quality

yl-ang committed:
Merge branch 'master' of https://github.com/yl-ang/ip

yl-ang committed:
Setup continuous Integration

yl-ang committed:
Update Junit test cases

Removing Storage test cases as it is difficult to test Storage without
proper stub and introducing additional test cases for ToDo and Unmark
command to improve unit testing of Duke.

Testing todo toStringForStorage method as well as Unmark's constructor
and isExit status.

This will ensure that the test cases closely test the program for bugs.

yl-ang committed:
Add Update command

yl-ang committed:
Improve Update Command

j4ck990 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,134 @@
+
+        } else {
+
+            // Event handling
+
+            if ((temp.length == 1) || (temp[1].equals("/at"))) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_EVENT_DESC.toString());
+            }
+
+            if (!data.contains("/at")) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_EVENT_DATE.toString());
+            }
+
+            String desc = extractDesc(data, "event ", "/at");
+            String date = extractTime(data, "/at ");
+            task = new Event(desc, date);
+        }
```
Would it be better to have event handling in another else if branch since its an if condition that can be easily specified? 
Perhaps it would help with less error-prone code quality.

j4ck990 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,92 @@
+
+            int i;
+            try {
+                i = Integer.parseInt(fullCommand.substring("unmark".length() + 1));
+            } catch (IndexOutOfBoundsException e) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_UNMARKED_INT.toString());
+            } catch (NumberFormatException e) {
+                throw new DukeException(ErrorString.ERROR_NOT_INT_UNMARKED.toString());
+            }
+
+            currCommand = new UnmarkCommand(i);
+
+        } else if (fullCommand.startsWith("delete ")) {
+
+            int i;
+
+            try {
+                i = Integer.parseInt(fullCommand.substring("delete".length() + 1));
+            } catch (IndexOutOfBoundsException e) {
+                throw new DukeException(ErrorString.ERROR_EMPTY_DELETE.toString());
+            } catch (NumberFormatException e) {
+                throw new DukeException(ErrorString.ERROR_NOT_NUMBER_STRING.toString());
+            }
+
+            currCommand = new DeleteCommand(i);
+        } else if (fullCommand.startsWith("todo ")) {
+            currCommand = new AddCommand(fullCommand, "todo");
+        } else if (fullCommand.startsWith("deadline ")) {
+            currCommand = new AddCommand(fullCommand, "deadline");
+        } else if (fullCommand.startsWith("event ")) {
+            currCommand = new AddCommand(fullCommand, "event");
+        } else if (fullCommand.startsWith("find ")) {
+            currCommand = new FindCommand(fullCommand);
+        } else if (fullCommand.startsWith("bye")) {
+            currCommand = new ExitCommand();
+        } else if (fullCommand.startsWith("update ")) {
+            currCommand = new UpdateCommand(fullCommand);
+        } else {
+            throw new DukeException(ErrorString.ERROR_INVALID_COMMAND.toString());
+        }
```
Perhaps using a switch case might be helpful for this use case?

j4ck990 commented on src/main/java/duke/exception/ErrorString.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @author yl-ang
+ */
+public enum ErrorString {
```
Very nice way of keeping all the error messages centralized. I like how it makes things easy to change error messages.

j4ck990 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @return Boolean if the command is exit.
+     */
+    public abstract boolean isExit();
```
I noticed how all your commands have their individual private isExit boolean variables. Perhaps a better way can be to have a protected/private isExit variable in Command class that can be accessed by all inheriting classes? 
Inheriting classes can then set the isExit variable through their constructors or super's constructor. 

Overall, helps with duplicate code. This way, isExit() abstract method here in Command can just be 
```java
public boolean isExit() {
    return this.isExit;
}
```

yl-ang replied:
Hi j4ck, great advice, will refactor my code to follow this! 👍

j4ck990 reviewed (commented):
Overall, very clear code structure. I like how neat and clear different logic sections are. 
Especially the way you handled you error messages' strings. d(>_< )

yl-ang committed:
Improve Parser via switch case

yl-ang committed:
Improve Update command

Package common extract methods from Update and AddCommand class to
into a new class Extract to reduce code duplication.

Allow update command to be flexible in the order of which date and
description is supplied.

For instance, user can do the following:
update 1 d/read ab3 code t/2022-02-14
update 1 t/2022-02-14 d/read ab3 code

This will allow update command to be less strict in updating valid
user input.

yl-ang committed:
Fix import statement ordering

yl-ang committed:
Add brief User Guide (UG)

yl-ang committed:
Set theme jekyll-theme-hacker

yl-ang committed:
Update screenshot format from JPG to png

yl-ang committed:
Correct typos in user guide

yl-ang committed:
Add missing list feature into user guide

yl-ang committed:
Update user guide

yl-ang committed:
Update Ui.png

yl-ang committed:
Update user guide

yl-ang committed:
Update user guide command summary table

yl-ang committed:
Format user guide command summary table

yl-ang committed:
Format user guide command summary table
